### DISCUSSION TITLE
Auto match columns

### DISCUSSION THREAD

**Discussion by Frameck** [konnco/filament-import#36] [Ideas]
It would be nice to add the ability to choose whether the import should automatically match the columns to the values, because if you have a lot of columns it takes a long time to manually associate them each time.
So if `Auto match columns` is `true` the form with selects doesn't appear and it matches automatically based on the name of the columns.

Something like this:

![Schermata 2022-10-05 alle 12 39 10](https://user-images.githubusercontent.com/77396783/194042061-26ecb653-7295-4eda-a0bc-f68201bc6c52.png)

**Comment by frankyso** [comment]
The feature has been added on 1.2.6 please have a  test :)

**Comment by Sharoek** [comment]
What is the method called ? cannot find it

**Comment by hunterd** [comment]
Please I would like to know too since I'm trying to set up this feature but I can't find out how.
I tried to extend ImportAction class and overwrite getFields method to add a ->default() parameter to the generated Select field but in the ->default() 's closure context I can't access the $uploadedFile's data.
If I did understand well that's how you construct the options from the spreadsheet in the ->options() method of the Select field but when I try to call the same data (in order to try to match the options label with the field name and then return the option value eg: "4" for the "lastname" because "lastname" is `<option value="4">lastname</option>`) I can't call the $uploadedFile?->getRealPath() as you do in the ->options() 's closure context.

```
class UsersImportAction extends ImportAction
{
    /**
     * @param  ImportField|Field  $field
     * @return Field
     */
    private function getFields(ImportField|Field $field): Field
    {
        if ($field instanceof Field) {
            return $field;
        }

        return Select::make($field->getName())
            ->label($field->getLabel())
            ->helperText($field->getHelperText())
            ->required($field->isRequired())
            ->placeholder($field->getPlaceholder())
            ->default(function(callable $get) use($field){
                $uploadedFile = last($get('file') ?? []);
                $filePath = is_string($uploadedFile) ? $uploadedFile : $uploadedFile?->getRealPath(); //getRealPath() never exists in this context

                if (count($this->cachedHeadingOptions) == 0) {
                    $cachedHeadingOptions = $this->toCollection($filePath)->first()?->first()?->toArray();
                }

                $cachedHeadingOptions; //I'd like to get the options to know which one matches the field name 

                
                return strval( array_search($field->getName(), $cachedHeadingOptions->map(fn ($option) => $option->value())) ); //$option->value is an hypotetic way to get the option value
            })
            ->options(options: function (callable $get) {
                $uploadedFile = last($get('file') ?? []);
                $filePath = is_string($uploadedFile) ? $uploadedFile : $uploadedFile?->getRealPath();

                if (count($this->cachedHeadingOptions) == 0) {
                    return $this->cachedHeadingOptions = $this->toCollection($filePath)->first()?->first()?->toArray(); //getRealPath() exists sometimes here (when the file field is filled I guess)
                }

                return $this->cachedHeadingOptions;
            });
    }
}
```

**Reply by frankyso** [reply]
Hi @hunterd, sorry I don't really understand what you mean, have you define the `file` upload field?

**Reply by hunterd** [reply]
Hi @frankyso,
sorry I'm not clear.
Yes the `file` upload field exists but is empty then I fill it with a CSV file.
I would like to get the list of CSV fields (as you do in the ->options() directive) or a way to know the keys/values you just put in the newly generated options in order to know which one to set as default in the ->default() directive.

In the end I'm trying to make myself an auto match column system since I don't find how to do so natively.
I'm using the version 1.2.6

**Comment by Saifallak** [comment]
use https://github.com/konnco/filament-import/pull/77
